### PR TITLE
Support cuDNN in F.dropout

### PR DESCRIPTION
**Pull request by bonprosoft**
This PR implements cuDNN support for dropout operation.
Its implementation depends on https://github.com/cupy/cupy/pull/479 .

### CODE REVIEW

**Comment by Crissman** [comment]
Please update status.

**Comment by bonprosoft** [comment]
@Crissman I apologize for my late reply.
I will do it this week. (I have to update my code like `get_random_state()` in `n_step_rnn`.)

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/backends/cuda.py [hunk 1]
```diff
@@ -637,3 +641,36 @@ def should_use_cudnn_tensor_core(dtype):
...
+
+# ------------------------------------------------------------------------------
+# cupy.cudnn utility
+# ------------------------------------------------------------------------------
+
+def get_cudnn_dropout_states():
+    if not cudnn_enabled:
+        raise RuntimeError('cuDNN is not enabled.')
+
+    thread_id = threading.get_ident()
```
Comment (okuta) [inline chainer/backends/cuda.py]: Please support Python2.7 . Python2.7 does not have `get_ident()`.
For example.
```
if hasatter(threading, 'get_ident'):
    thread_id = threading.get_ident()
else:
    thread_id = threading.current_thread().ident
```

File: chainer/backends/cuda.py [hunk 2]
```diff
@@ -637,3 +641,36 @@ def should_use_cudnn_tensor_core(dtype):
...
+    if seed is None:
+        try:
+            seed_str = binascii.hexlify(os.urandom(8))
+            seed = numpy.uint64(int(seed_str, 16))
+        except NotImplementedError:
+            seed = numpy.uint64(time.clock() * 1000000)
+    else:
+        seed = numpy.uint64(seed)
+
+    seed = seed + numpy.uint64(states_id)
```
Comment (okuta) [inline chainer/backends/cuda.py]: `seed += ` is better.

**Comment by okuta** [comment]
rel #4134

File: chainer/backends/cuda.py (hunk 1 above)
Comment (bonprosoft) [reply chainer/backends/cuda.py]: According to documents, python 2.7 and 3.x have almost same interfaces for `threading.Thread`.
We can also obtain this object by calling `threading.current_thread()`.

https://docs.python.org/3/library/threading.html#threading.Thread.ident
https://docs.python.org/2.7/library/threading.html#threading.Thread.ident

So I use the following code for obtaining a thread id in both Python 2.7 and 3.x:
```python
thread_id = threading.current_thread().ident
```

**Comment by bonprosoft** [comment]
Thank you for your review. I fixed my code.
Could you check it again, please?

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
OK, but `dropout.py` is modified. Cloud you resolve conflicts?

File: chainer/functions/noise/dropout.py [hunk 3]
```diff
@@ -33,13 +37,33 @@ def forward(self, x):
...
-                flag = xp.random.rand(*x[0].shape) >= self.dropout_ratio
-                self.mask = scale * flag
+            flag = numpy.random.rand(*x[0].shape) >= self.dropout_ratio
+            self.mask = scale * flag
+            y = x[0] * self.mask
+        return y,
+
+    def forward_gpu(self, x):
+        if (chainer.should_use_cudnn('==always', 5000) and
+                x[0].flags.c_contiguous):
```
Comment (bonprosoft) [inline chainer/functions/noise/dropout.py]: Should I change it to:
```python
if (chainer.should_use_cudnn('==always', 5000) and x[0].flags.c_contiguous and self.mask is None):
```

It seems that now dropout function has `mask` argument in its constructor for making results deterministic, but cuDNN does not support `mask` (same as iDeep).

**Comment by bonprosoft** [comment]
@okuta Rebased. I also write a comment. Could you check it, please?

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**